Lib/glyphsLib/builder/bracket_layers.py

```diff
@@ -4,117 +4,55 @@
-                max_crossovers.add(bracket_max)
-        # reverse and non-reverse bracket layers with overlapping ranges are
-        # tricky to implement as DS rules. They are relatively unlikely, and
-        # can usually be rewritten so that they do not overlap. For laziness/
-        # simplicity, where we simply warn that output may not be as expected.
-        invalid_locs = [
-            (mx, mn)
-            for mx, mn in zip(sorted(max_crossovers), sorted(min_crossovers))
-            if mx > mn
-        ]
-        if invalid_locs:
-            self.logger.warning(
-                "Bracket layers for glyph '%s' have overlapping ranges: %s",
-                glyph_name,
-                ", ".join("]{}] > [{}]".format(*values) for values in invalid_locs),
+        # It'd be nice to use tag for both glyph names and designspace
+        # rules, but given designspace refers to axes by name, our hand
+        # is forced. (We don't use axis names in glyph names because they
+        # might have spaces in them.)
+        box_with_name = {tag_to_name[k]: v for k, v in box_with_tag.items()}
+        bracket_layer_map[glyph_name][util.freezedict(box_with_tag)].append(layer)
+        rules.append(
+            (
+                [box_with_name],
+                {glyph_name: _bracket_glyph_name(glyph_name, box_with_tag)},
             )
-        max_crossovers = list(sorted(max_crossovers))
-        if bracket_axis_min not in max_crossovers:
-            max_crossovers = [bracket_axis_min] + max_crossovers
-        for crossover_min, crossover_max in util.pairwise(max_crossovers):
-            max_rule_bucket[(int(crossover_min), int(crossover_max))].append(glyph_name)
-        min_crossovers = list(sorted(min_crossovers))
-        if bracket_axis_max not in min_crossovers:
-            min_crossovers = min_crossovers + [bracket_axis_max]
-        for crossover_min, crossover_max in util.pairwise(min_crossovers):
-            min_rule_bucket[(int(crossover_min), int(crossover_max))].append(glyph_name)
+        )
 
-    # Generate rules for the bracket layers.
-    for reverse, rule_bucket in ((True, max_rule_bucket), (False, min_rule_bucket)):
-        for (axis_range_min, axis_range_max), glyph_names in sorted(
-            rule_bucket.items()
-        ):
-            rule = _make_designspace_rule(
-                glyph_names,
-                bracket_axis.name,
-                axis_range_min,
-                axis_range_max,
-                reverse,
-            )
-            self._designspace.addRule(rule)
+    # overlayFeatureVariations does exactly what we need in terms of
+    # splitting rules into non-overlapping boxes and consolidating
+    # multiple substitutions into a single rule.
+    for rule in overlayFeatureVariations(rules):
+        box, mappings = rule
+        # Reduce the mappings to a single mapping
+        mapping = {}
+        for this_mapping in mappings:
+            mapping.update(this_mapping)
```

Comment:
you could also use `collections.ChainMap` if you like, I believe it does the same as iterating and updating
https://docs.python.org/3/library/collections.html#collections.ChainMap

```python
mapping = dict(ChainMap(*mappings))
```

Reply:
Nice idea. I looked into this, and found it applied the chaining the wrong way around (last to first). Then I tried `mapping = dict(ChainMap(*reversed(mappings)))` which works, but I felt it was Just Too Much Magic and what I originally wrote is considerably easier to follow.